### DISCUSSION TITLE
Whether setting security level to zero in FIPS mode should be allowed or not ?

### DISCUSSION THREAD

**Discussion by ashwanikadian** [openssl/openssl#24983] [Q&A] [answered]
I was trying to test anonymous ciphers with OpenSSL 3.0.14 server and OpenSSL 3.0.14 client. I observed that as mentioned in OpenSSL issue 18771 (https://github.com/openssl/openssl/issues/18771), we need to set security level to zero to get the anonymous ciphers working.

So, I tried setting security level to zero on both server and client sides and I was able to use anonymous ciphers successfully. But then I tried to do the same test in FIPS mode. In FIPS mode also, I was able to set the security level to zero and use the anonymous ciphers successfully. 

The page https://www.openssl.org/docs/man3.0/man3/SSL_CTX_set_security_level.html tells that everything is allowed at security level zero, including legacy algorithms and lower key sizes, which are not FIPS compliant. 

So, I wanted to confirm:  Is setting security level to zero in FIPS mode safe or not? I mean, will it have any security concern in FIPS mode in any other functional area? Should setting security level to zero be allowed in FIPS mode or not? Will my application still be FIPS compliant, if I set the security level to zero?

**Comment by nhorman** [comment]
I believe the security level is independent of any FIPS validation.  That is to say if you have the fips provider properly installed and configured, and set the security level to 0.  libssl will allow smaller key sizes, but fips may still preform validation on those keys, and reject them if they are not fips approved.

@mattcaswell can you please confirm?

**Comment by tom-cosgrove-arm** [comment]
(aside: should this be a discussion rather than an issue?)

**Reply by paulidale** [reply]
I think so

**Comment by paulidale** [comment]
This is the behaviour I'd expect.  Security level isn't related to FIPS.

**Comment by mattcaswell** [comment, marked as answer]
> I believe the security level is independent of any FIPS validation. That is to say if you have the fips provider properly installed and configured, and set the security level to 0. libssl will allow smaller key sizes, but fips may still preform validation on those keys, and reject them if they are not fips approved.

Right - security level is independent. The security level is a libssl restriction. It will restrict the things that libssl can do (including the algorithms it will attempt to use). Using the FIPS provider (and not having the default provider available) will restrict the algorithms/key sizes available to libcrypto. Since libssl uses libcrypto for its crypto, what you end up with as usable is the intersection between these two things.

If you set the security level to 0 and have the fips module loaded this has no impact on your conformance to FIPS as long as you don't also have algorithms from another provider available like "legacy" or "default".

**Comment by ashwanikadian** [comment]
Thanks a lot for the quick confirmation. This really helps :)
